### PR TITLE
Feat/scatter

### PR DESCRIPTION
**Pull request by Tanguylo**
New implementation of Scatter plot with Frame as parent:
image

### CODE REVIEW

File: src/multiplots.ts [hunk 1]
```diff
@@ -1758,6 +1779,35 @@ export class MultiplePlots {
...
+          })
+        }
+      })
+    }
+
+    public refreshRubberBands() {
+      if (!this.rubberBands) this.initRubberBands();
+      this.objectList.forEach(plot => {
+        if (plot instanceof BasePlot) {
+          plot.axes.forEach(axis => {
```
Comment (GhislainJ) [inline src/multiplots.ts:1799]: I think this can be an axis method as well

File: src/multiplots.ts [hunk 2]
```diff
@@ -2035,34 +2084,38 @@ export class MultiplotCom {
     }
 
     public static frame_to_frame_communication(frame1, frame2) {
-      if (frame1.axes[0].name !== frame2.axes[0].name) return;
-      frame2.axes[0].rubberBand = frame1.axes[0].rubberBand;
-      frame2.axes[1].rubberBand = frame1.axes[1].rubberBand;
+      frame2.axes.forEach(axisFrame2 => {
+        frame1.axes.forEach(axisFrame1 => {
+          if (axisFrame1.name == axisFrame2.name && axisFrame1.name != 'number') {
```
Comment (GhislainJ) [inline src/multiplots.ts:2089]: What does the "number" named frame stand for ?

File: src/plot-data.ts [hunk 3]
```diff
@@ -2273,6 +2273,7 @@ export abstract class PlotData {
       var canvas = document.getElementById(this.canvas_id);
 
       canvas.addEventListener('mousedown', e => {
+        console.log(e.offsetX, e.offsetY)
```
Comment (GhislainJ) [inline src/plot-data.ts:2276]: Leftover log ?

File: src/utils.ts [hunk 4]
```diff
@@ -1328,31 +1331,45 @@ export class newShape {
...
+  protected _tooltipOrigin: Vertex;
+  protected _tooltipMap = new Map<string, any>();
   protected readonly TOOLTIP_SURFACE: SurfaceStyle = new SurfaceStyle(string_to_hex("lightgrey"), 0.5, null);
   protected readonly TOOLTIP_TEXT_STYLE: TextStyle = new TextStyle(string_to_hex("black"), 14, "Calibri");
   constructor() {};
 
-  protected computeTooltipOrigin(contextMatrix: DOMMatrix): Vertex {
-    throw new Error(`Method computeTooltipOrigin not implemented for ${this.constructor.name}`)
-  }
+  get tooltipOrigin(): Vertex { return this._tooltipOrigin };
```
Comment (GhislainJ) [inline src/utils.ts:1340]: Are these getters/setters necessary ?

File: src/utils.ts [hunk 5]
```diff
@@ -2151,6 +2164,14 @@ export class newAxis {
 
   get minValue(): number { return this._minValue };
 
+  set initMaxValue(value: number) { this._initMaxValue = value };
```
Comment (GhislainJ) [inline src/utils.ts:2167]: Unnecessary getters/setters ?

File: src/utils.ts [hunk 6]
```diff
@@ -2224,12 +2245,10 @@ export class newAxis {
...
   }
 
-  public normedValue(value: number): number {
-    return value / this.interval
-  }
+  public normedValue(value: number): number { return value / this.interval }
 
   private computeMinMax(vector: any[]): number[] {
-    if (this.isDiscrete) { return [0, this.labels.length] };
+    if (this.isDiscrete) { return [0, this.labels.length - 1] };
```
Comment (GhislainJ) [inline src/utils.ts:2251]: If you fancy it, you can remove redundant brackets here (`if (this.isDiscrete) return [0, this.labels.length - 1];`). That is cleaner for one-liner ifs

File: src/subplots.ts [hunk 7]
```diff
@@ -1489,19 +1490,20 @@ export class BasePlot extends PlotData {
...
   private _axisStyle = new Map<string, any>([['strokeStyle', 'hsl(0, 0%, 31%)']]);
-  private nSamples: number;
+  public nSamples: number;
 
   readonly features: Map<string, any[]>;
   readonly MAX_PRINTED_NUMBERS = 16;
   readonly TRL_THRESHOLD = 20;
   constructor(
-    public data: any,
+    data: any,
```
Comment (GhislainJ) [inline src/subplots.ts:1506]: Why is this necessary to remove the public ?

File: src/subplots.ts [hunk 8]
```diff
@@ -1634,30 +1657,34 @@ export class BasePlot extends PlotData {
...
+  public mouseDown(canvasMouse: Vertex, frameMouse: Vertex, absoluteMouse: Vertex): [Vertex, Vertex, any] {
     let clickedObject: any;
     this.fixedObjects.forEach(object => {if (object.isHovered) {clickedObject = object}})
-    this.movingObjects.forEach(object => {if (object.isHovered) {clickedObject = object}})
+    this.absoluteObjects.forEach(object => {if (object.isHovered) {clickedObject = object}})
+    this.relativeObjects.forEach(object => {if (object.isHovered) {clickedObject = object}})
     if (this.fixedObjects.indexOf(clickedObject) != -1) {clickedObject.mouseDown(canvasMouse)}
-    if (this.movingObjects.indexOf(clickedObject) != -1) {clickedObject.mouseDown(frameMouse)}
+    if (this.absoluteObjects.indexOf(clickedObject) != -1) {clickedObject.mouseDown(absoluteMouse)}
+    if (this.relativeObjects.indexOf(clickedObject) != -1) {clickedObject.mouseDown(frameMouse)}
```
Comment (GhislainJ) [inline src/subplots.ts:1678]: I am puzzled by the fact that we often need to do exactly the same stuff (or almost) to absolute and relativeObjects. This often mark a structure that should be OO.

Relative and Absolute objects would inherit from an object and specify their own way of doing these

File: src/subplots.ts [hunk 9]
```diff
@@ -1694,19 +1720,22 @@ export class BasePlot extends PlotData {
...
+        [canvasDown, frameDown, clickedObject] = this.mouseDown(canvasMouse, frameMouse, absoluteMouse);
+        if (clickedObject instanceof newAxis) this.is_drawing_rubber_band = true
+        else this.is_drawing_rubber_band = false
         isDrawing = true;
       });
 
       canvas.addEventListener('mouseup', e => {
         canvas.style.cursor = 'default';
-        this.mouseUp(canvasMouse, frameMouse, canvasDown, ctrlKey);
+        console.log(e.offsetX, e.offsetY)
```
Comment (GhislainJ) [inline src/subplots.ts:1731]: Leftover log ?

File: src/subplots.ts [hunk 10]
```diff
@@ -1694,19 +1720,22 @@ export class BasePlot extends PlotData {
...
+        this.mouseUp(canvasMouse, frameMouse, absoluteMouse, canvasDown, ctrlKey);
         if (clickedObject) {clickedObject.mouseUp()};
-        isDrawing = false;
         this.draw();
         this.axes.forEach(axis => {axis.saveLocation()});
         this.translation = new Vertex(0, 0);
-        this.is_drawing_rubber_band = false;
+        clickedObject = undefined;
```
Comment (GhislainJ) [inline src/subplots.ts:1737]: Dangerous undefined assignment. Consider assigning to null and handling objects when null. undefined is a marker that something went wrong, a variable has not been passed of assigned, but developers should not set something as undefined. Weirdly enough, undefined is a different concept from "I don't want my variable to have a value"

File: src/subplots.ts [hunk 11]
```diff
@@ -2045,3 +2075,78 @@ export class Histogram extends Frame {
...
+    public height: number,
+    public buttons_ON: boolean,
+    public X: number,
+    public Y: number,
+    public canvas_id: string,
+    public is_in_multiplot: boolean = false
+    ) {
+      super(data, width, height, buttons_ON, X, Y, canvas_id, is_in_multiplot);
+      if (!data.tooltip) {this.tooltipAttr = Array.from(this.features.keys()) }
+      else this.tooltipAttr = data.tooltip.attribute;
```
Comment (GhislainJ) [inline src/subplots.ts:2097]: Now I know this isn't tooltipAttrition ;)

File: src/subplots.ts [hunk 12]
```diff
@@ -2045,3 +2075,78 @@ export class Histogram extends Frame {
...
+
+  private computePoints(): newPoint2D[] {
+    const numericVectorX = this.axes[0].stringsToValues(this.features.get(this.xFeature));
+    const numericVectorY = this.axes[1].stringsToValues(this.features.get(this.yFeature));
+    const points: newPoint2D[] = [];
+    const pointsHTMatrix = this.getPointsHTMatrix(numericVectorX, numericVectorY);
+    const coordsMatrix = this.projectPoints(pointsHTMatrix);
+    for (let index = 0 ; index < numericVectorX.length ; index++) {
+      let newPoint = new newPoint2D(coordsMatrix.get(0, index),  coordsMatrix.get(1, index), 5, "circle", undefined, "hsl(203, 90%, 85%)");
+      if (this.hoveredIndices[index]) {newPoint.isHovered = true};
```
Comment (GhislainJ) [inline src/subplots.ts:2108]: Why not `newPoint.isSomethinged = this.somethingedIndices[index]` ?

File: src/subplots.ts [hunk 13]
```diff
@@ -2045,3 +2075,78 @@ export class Histogram extends Frame {
...
+      if (this.clickedIndices[index]) {newPoint.isClicked = true};
+      if (this.selectedIndices[index]) {newPoint.isSelected = true};
+
+      const inCanvasX = numericVectorX[index] < this.axes[0].maxValue && numericVectorX[index] > this.axes[0].minValue;
+      const inCanvasY = numericVectorY[index] < this.axes[1].maxValue && numericVectorY[index] > this.axes[1].minValue;
+      if (inCanvasX && inCanvasY) {
+        if (newPoint.isClicked) this.tooltipAttr.forEach(attr => newPoint.tooltipMap.set(attr, this.features.get(attr)[index]));
```
Comment (GhislainJ) [inline src/subplots.ts:2115]: And just now, I realise that tooltipAttr is an string array, and should be called tooltipAttribute**s** :) 

DO NOT EVEN THINK ABOUT NAMING IT `tooltipAttrs` !

File: src/subplots.ts [hunk 14]
```diff
@@ -2045,3 +2075,78 @@ export class Histogram extends Frame {
     return [mouse3X, mouse3Y];
   }
 }
+
+
+export class newScatter extends Frame {
+  public points: newPoint2D[] = [];
+  public tooltipAttr: string[];
```
Comment (GhislainJ) [inline src/subplots.ts:2082]: Consider fully naming variables. This could be tooltipAttraction or tooltipAttrition. (i know, this is a bit far-fetched, but still, I think it is important to properly name variables)

File: src/subplots.ts [hunk 15]
```diff
@@ -2045,3 +2075,78 @@ export class Histogram extends Frame {
...
+        if (newPoint.isClicked) this.tooltipAttr.forEach(attr => newPoint.tooltipMap.set(attr, this.features.get(attr)[index]));
+        newPoint.draw(this.context_show);
+      }
+      points.push(newPoint);
+    }
+    return points
+  }
+
+  private getPointsHTMatrix(vectorX: number[], vectorY: number[]) {
+    return new matrices.Matrix([vectorX, vectorY, Array.from(Array(this.nSamples), () => 1)])
```
Comment (GhislainJ) [inline src/subplots.ts:2124]: What does this line do ? I am a bit puzzled by the amount of arguments given

File: src/subplots.ts [hunk 16]
```diff
@@ -2045,3 +2075,78 @@ export class Histogram extends Frame {
...
+    const relativeMatrix = new matrices.Matrix([
+      [this.relativeMatrix.a, this.relativeMatrix.c, this.relativeMatrix.e],
+      [this.relativeMatrix.b, this.relativeMatrix.d, this.relativeMatrix.f],
+      [0, 0, 1]
+    ])
+    return relativeMatrix.mmul(pointsMatrix)
+  }
+
+  public stateUpdate(context: CanvasRenderingContext2D, objects: any[], mouseCoords: Vertex, stateName: string, keepState: boolean, invertState: boolean): void {
+    // TODO: Fast and dirty implementation. Needs to be rethought ?
```
Comment (GhislainJ) [inline src/subplots.ts:2137]: I actually am struggling to keep track of what is going on here

**Review by GhislainJ — CHANGES_REQUESTED** [review]

File: cypress/e2e/histogram.cy.ts [hunk 17]
```diff
@@ -68,13 +68,16 @@ describe('HISTOGRAM CANVAS', function () {
   it("should hover/click on bar", function () {
     cy.window().then(win => {
       const histogram = win.eval('plot_data') as Histogram;
-      [canvasMouse, frameMouse] = histogram.projectMouse({"offsetX": 278, "offsetY": 530} as MouseEvent);
-      histogram.mouseMove(canvasMouse, frameMouse);
+      [canvasMouse, frameMouse, mouseCoords] = histogram.projectMouse({"offsetX": 348, "offsetY": 399} as MouseEvent);
+      histogram.mouseMove(canvasMouse, frameMouse, mouseCoords);
+      console.log(histogram.hoveredIndices)
```
[the remark this answers is not in the document]
Comment (Tanguylo) [reply cypress/e2e/histogram.cy.ts:73]: Removed

File: src/multiplots.ts [hunk 18]
```diff
@@ -129,6 +134,7 @@ export class MultiplePlots {
         this.clean_view();
         this.store_dimensions();
       }
+      this.refreshRubberBands();
         // this.save_canvas();
```
[the remark this answers is not in the document]
Comment (Tanguylo) [reply src/multiplots.ts:138]: Not my code

File: src/multiplots.ts [hunk 19]
```diff
@@ -571,11 +577,15 @@ export class MultiplePlots {
...
         this.display_order = List.move_elements(old_index, this.display_order.length - 1, this.display_order);
       }
-
-      this.objectList.forEach((plot, index) => {
-        if (List.is_include(index, this.to_display_plots)) {
+      this.objectList.forEach((plot, pIndex) => {
+        if (List.is_include(pIndex, this.to_display_plots)) {
           if (plot.type_ == 'parallelplot') { plot.refresh_axis_coords() }
-          // plot.mouse_interaction(plot.isParallelPlot);
+          if (plot instanceof BasePlot) {
```
[the remark this answers is not in the document]
Comment (Tanguylo) [reply src/multiplots.ts:583]: It is but not in case of parallelplot, just above

File: src/multiplots.ts [hunk 20]
```diff
@@ -571,11 +577,15 @@ export class MultiplePlots {
...
+            plot.selectedIndices = Array.from(Array(plot.selectedIndices.length), (value, vIndex) => {
+              if (this.dep_selected_points_index.includes(vIndex)) return true
```
[the remark this answers is not in the document]
Comment (Tanguylo) [reply src/multiplots.ts:585]: Done

File: src/multiplots.ts [hunk 21]
```diff
@@ -852,12 +862,19 @@ export class MultiplePlots {
         } else if ((obj.type_ === 'parallelplot') && !List.isListOfEmptyList(obj.rubber_bands)) {
           bool = true;
           this.dep_selected_points_index = List.listIntersection(this.dep_selected_points_index, obj.pp_selected_index);
+        } else if (obj instanceof BasePlot) {
```
[the remark this answers is not in the document]
Comment (Tanguylo) [reply src/multiplots.ts:865]: We are not anymore in next branches

File: src/multiplots.ts [hunk 22]
```diff
@@ -852,12 +862,19 @@ export class MultiplePlots {
...
+          obj.axes.forEach(axis => {
+            if (axis.rubberBand.length != 0) {
+              bool = true;
```
[the remark this answers is not in the document]
Comment (Tanguylo) [reply src/multiplots.ts:868]: Done

File: src/multiplots.ts [hunk 23]
```diff
@@ -852,12 +862,19 @@ export class MultiplePlots {
...
+              const boolIndex = (obj as BasePlot).updateSelected(axis);
+              let selectedIndices = [];
+              boolIndex.forEach((bool, bIndex) => { if (bool) selectedIndices.push(bIndex) })
```
[the remark this answers is not in the document]
Comment (Tanguylo) [reply src/multiplots.ts:871]: Next branches

File: src/multiplots.ts [hunk 24]
```diff
@@ -884,7 +901,7 @@ export class MultiplePlots {
     }
 
     resetAllObjects(): void {
-      this.dep_selected_points_index = [];
+      // this.dep_selected_points_index = [];
```
[the remark this answers is not in the document]
Comment (Tanguylo) [reply src/multiplots.ts:904]: Some changes have been done

File: src/multiplots.ts [hunk 25]
```diff
@@ -1338,17 +1355,21 @@ export class MultiplePlots {
               subplot.select_on_click.push(scatterPoint)
             }
           })
-          subplot.refresh_selected_point_index();
+          // subplot.refresh_selected_point_index();
```
[the remark this answers is not in the document]
Comment (Tanguylo) [reply src/multiplots.ts:1358]: Removed

File: src/multiplots.ts [hunk 26]
```diff
@@ -1758,6 +1779,35 @@ export class MultiplePlots {
       return click_on_manip_button || click_on_selectDep_button || click_on_view || click_on_export;
     }
 
+    public initRubberBands() {
+      this.rubberBands = new Map<string, RubberBand>();
+      this.objectList.forEach(plot => {
+        if (plot instanceof BasePlot) {
+          plot.axes.forEach(axis => { this.rubberBands.set(axis.name, new RubberBand(axis.name, 0, 0, axis.isVertical)) })
```
[the remark this answers is not in the document]
Comment (Tanguylo) [reply src/multiplots.ts:1786]: Done, it is a method of rubberband

File: src/multiplots.ts [hunk 27]
```diff
@@ -1758,6 +1779,35 @@ export class MultiplePlots {
...
+        } else if (plot instanceof ParallelPlot) {
```
[the remark this answers is not in the document]
Comment (Tanguylo) [reply src/multiplots.ts:1787]: It is. In further releases, we will just write plot.function without asking for its type because it is a non sense in POO

File: src/multiplots.ts (hunk 2 above)
Comment (Tanguylo) [reply src/multiplots.ts:2089]: changed to currentFrame and ohterFrame

File: src/plot-data.ts (hunk 3 above)
Comment (Tanguylo) [reply src/plot-data.ts:2276]: Removed
I like to keep it to know where my mouse is

File: src/utils.ts (hunk 4 above)
Comment (Tanguylo) [reply src/utils.ts:1340]: Removed

File: src/utils.ts (hunk 5 above)
Comment (Tanguylo) [reply src/utils.ts:2167]: Yes

File: src/utils.ts (hunk 6 above)
Comment (Tanguylo) [reply src/utils.ts:2251]: Done

File: src/subplots.ts (hunk 7 above)
Comment (Tanguylo) [reply src/subplots.ts:1506]: I think data is not an attribute of plot, i removed it from all plots. I cannot for PlotData object yet so it is still an instance attribute, but should not.It should just be used to instantiate plot and then thrown

File: src/subplots.ts (hunk 15 above)
Comment (Tanguylo) [reply src/subplots.ts:2124]: Removed in next branches

File: src/subplots.ts (hunk 8 above)
Comment (Tanguylo) [reply src/subplots.ts:1678]: You're right and i've done it

Comment (Tanguylo) [reply src/subplots.ts]: Maybe not in an optimized way but it introduces a fine way to do what has to be done

File: src/subplots.ts (hunk 9 above)
Comment (Tanguylo) [reply src/subplots.ts]: Removed

File: src/subplots.ts (hunk 10 above)
Comment (Tanguylo) [reply src/subplots.ts:1731]: Done

File: src/subplots.ts (hunk 11 above)
Comment (Tanguylo) [reply src/subplots.ts:2084]: Replaced

File: src/subplots.ts (hunk 12 above)
Comment (Tanguylo) [reply src/subplots.ts:2095]: Done

File: src/subplots.ts (hunk 13 above)
Comment (Tanguylo) [reply src/subplots.ts:2102]: :)

File: src/subplots.ts (hunk 14 above)
Comment (Tanguylo) [reply src/subplots.ts:2069]: ok

File: src/subplots.ts (hunk 16 above)
Comment (Tanguylo) [reply src/subplots.ts]: It has been removed with the above modifications

File: src/multiplots.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (GhislainJ) [reply src/multiplots.ts:138]: In this case, it can be removed as it won't be useless again I think.
Globally, every leftover log or comment from the past can be removed

File: src/multiplots.ts (hunk 2 above)
Comment (GhislainJ) [reply src/multiplots.ts]: Ok with the change (could even be `frame` and `other`), but what usecase denotes this check : `axisFrame1.name != 'number'` ?

File: src/multiplots.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (Tanguylo) [reply src/multiplots.ts:138]: Done

File: src/multiplots.ts (hunk 2 above)
Comment (Tanguylo) [reply src/multiplots.ts]: This denotes the case of Histogram number axis which must not be changed with it. I agree this is not the best way but it is the simplest and will change with multiplot refactor

**Review by GhislainJ — APPROVED** [review]
